Raising warnings when submitting team response with assignees who aren't in organization

woojiahao opened the pull request:
### Summary:
Fixes #1262 

### Changes Made:
* Custom parsing of `updateIssue` errors to detect when assignees are invalid and providing custom error messages
* Inverts the flow of creating a team response (current flow still creates the comment even if the assignee information is wrong, but this causes the issue to be falsely flagged as "responded"

### Proposed Commit Message:
```
Resolve issue of allocating invalid assignees

Currently, when a user attempts to submit a team response with an 
invalid assignee (defined as an assignee who has not joined the 
organization), the team response is processed (by adding the team 
response GitHub comment) before the assignee checks are made. The error 
message for an invalid assignee is also vague, only stating "Validation
Failed", with no additional context.

This causes the issue to be erroneously classified as “Responded” when
it has not actually been properly responded to as the “assignee” field 
would be left empty. The vague invalid assignee message also means users
will not know what to do to fix the issue.

This issue can be resolved by inverting the logic of team responses, 
ensuring the assignee validity is checked before making the team 
response comment. If the assignee is invalid, then the team response 
comment is never created, thus avoiding the erroneous classification. 
Also provided custom error parsing for invalid assignees to display 
proper reasons and actions for the validation failure for users to 
rectify the issue.

A proper diagram of the inversion of logic can be found in the pull
request:
https://github.com/CATcher-org/CATcher/pull/1264#issuecomment-2022201156
```

woojiahao commented:
For the second fix, the primary issue is that the comment is added before the assignee issue could be detected which falsely sets the issue to be "responded" when the assignee field is actually blank.

Inverting the order would resolve this issue logically, but the current `updateIssue` is dependent on the updated GitHub comments which means that the inherent behavior is tied to the old ordering (create comment -> assign issue metadata) than the new one.

I'm currently investigating ways to fix this issue, likely to use an optional parameter in `updateIssue` to minimize overall code changes but an alternative is to create a "shadow" copy of `updateIssue` that does not perform the automatic parsing from `GitHubIssue` -> `Issue` until the final operation is completed.

woojiahao commented:
For the sake of clarity, the following is the original flow of creating a team response

```mermaid
sequenceDiagram
  participant Caller
  participant i as IssueService
  participant g as GitHubService
  Caller ->> i : createTeamResponse(Issue)
  i ->> g : createIssueComment
  g -->> i : GitHubComment
  i ->> i : Update Issue githubComments field with latest comments
  i ->> i : updateIssue
  i ->> g : updateIssue
  alt valid fields
  g -->> i : GitHubIssue
  i ->> i : createIssueModel(GitHubIssue)
  else else
  g -->> i : Error
  i ->> i : Parse error
  end
  i -->> Caller : Complete
```

The new flow proposed does the following

```mermaid
sequenceDiagram
  participant Caller
  participant i as IssueService
  participant g as GitHubService
  Caller ->> i : createTeamResponse(Issue)
  i ->> i : updateIssuePure
  i ->> g : updateIssue
  alt valid fields
  g -->> i : GitHubIssue
  i ->> i : createIssueModel(GitHubIssue)
  i ->> g : createIssueComment
  g -->> i : GitHubComment
  i ->> i : Update Issue githubComments field with latest comments
  i -->> Caller : Complete
  else else
  g -->> i : Error
  i ->> i : Parse error
  i -->> Caller : Terminate operation
  end
```

luminousleek commented on src/app/shared/view-issue/new-team-response/new-team-response.component.ts:
```diff
@@ -108,9 +108,9 @@ export class NewTeamResponseComponent implements OnInit, OnDestroy {
     this.isSafeToSubmit()
       .pipe(
-        mergeMap((isSaveToSubmit: boolean) => {
+        mergeMap((isSafeToSubmit: boolean) => {
```
Thanks for the catch!

luminousleek commented on src/app/core/services/issue.service.ts:
```diff
@@ -144,8 +144,24 @@ export class IssueService {
+   * @param issue current issue model
+   * @returns GitHubIssue from the API request
+   */
+  updateIssuePure(issue: Issue): Observable<GithubIssue> {
```
Seems like this is duplicating the first part of the `updateIssue` method. What do you think about renaming this method as `updateGithubIssue` and then refactoring `updateIssue` to something like
```javascript
updateIssue(issue: Issue): Observable<Issue> {
  return this.updateGithubIssue
    .pipe(
      map((response: GithubIssue) => {
        response.comments = issue.githubComments;
        return this.createIssueModel(response);
      })
    );
}
```

woojiahao replied:
Yup! That makes sense, didn't really know what to name it haha, but will do!

luminousleek reviewed (changes requested):
Thanks for all your hard work investigating and fixing the issue! The inversion of logic makes sense. Wondering if we should do a refactor of the Issue Service to standardise this.

Just a minor comment so that we can follow the DRY principle. Also, please follow the commit message guidelines at [this link](https://se-education.org/guides/conventions/git.html)

woojiahao commented:
@luminousleek Thank you for the review :) I have updated my commit message and addressed the comments!

Re refactoring the issues service: it does appear that a final step for several other methods is to perform an effective DTO mapping from `GithubIssue` to `Issue`. Perhaps the team might want to consider a proper logical split between the logic that retrieves data from the Github API and the logic that performs the DTO mapping. I feel that this can help to make the code a lot clearer. It also helps to create more atomic operations for issue querying with the Github API.

luminousleek reviewed (approved):
LGTM! Thanks for both raising the issue and spending the time to set up CATcher to fix it. 

Your refactor idea to split the logic for the GitHub API calls and the DTO mapping makes sense - if you happen to have too much time maybe you could submit an issue for it.
